### PR TITLE
Keep projects open on tsserver via OpenExternalProject

### LINKED ISSUE #142: Language services require at least one open file
The language services provided by tsserver require at least one file to be open among the ones belonging to the scope of a _tsconfig.json_.
A file is open if an `open` was sent, but no corresponding `close` command). With the current typescript.java architecture, this means that a TypeScript Editor must be open.

If a _tsconfig.json_ has no file open, no language support will be available for it. This notably includes `compileOnSaveAffectedFileList`, but also simpler commands like `projectInfo` do not work.

In addition to missed compilation, this can cause all sorts of problems in Eclipse: for example when refreshing external changes, doing a Find-Replace, moving/renaming files or checking out from Git.

### Steps to reproduce

In a project with _compile on save_ enabled:
1. close all files,
2. modify a .ts file out of Eclipse,
3. refresh the file in Eclipse,

The file itself is _not_ recompiled.

### PR DESCRIPTION
**Pull request by lorenzodallavecchia**
This change makes tsserver aware of existing Eclipse projects, so that it does not forget about project state when all files are closed.
Closes #142.

### CODE REVIEW

File: core/ts.core/src/ts/client/ITypeScriptServiceClient.java [hunk 1]
```diff
@@ -91,8 +92,8 @@
 	 * @param insertString
 	 * @throws TypeScriptException
 	 */
-	void changeFile(String fileName, int line, int offset, int endLine, int endOffset, String insertString)
-			throws TypeScriptException;
+	void changeFile(String fileName, int line, int offset, int endLine, int endOffset,
```
Comment (angelozerr) [inline core/ts.core/src/ts/client/ITypeScriptServiceClient.java]: Please use the same Java Format preferences (Eclipse [built-in].

File: core/ts.core/src/ts/resources/TypeScriptProject.java [hunk 2]
```diff
@@ -134,13 +150,20 @@ public final ITypeScriptServiceClient getClient() throws TypeScriptException {
 					this.client = createServiceClient(getProjectDir());
 					copyListeners();
 					onCreateClient(client);
+					// determine root files and project name
+					String projectName = projectDir.getCanonicalPath();
+					List<String> rootFiles = new ArrayList<>();
+					for (String tsconfigFilePath : getTsconfigFilePaths()) {
+						rootFiles.add(new File(projectDir, tsconfigFilePath).getCanonicalPath());
```
Comment (angelozerr) [inline core/ts.core/src/ts/resources/TypeScriptProject.java]: Please use FileUtils.getPath(new File(projectDir, tsconfigFilePath)) to normalize path.

File: eclipse/ts.eclipse.ide.core/src/ts/eclipse/ide/internal/core/resources/IDETypeScriptProject.java [hunk 3]
```diff
@@ -352,6 +353,17 @@ public void disposeBuildPath() {
 	}
 
 	@Override
+	protected List<String> getTsconfigFilePaths() {
+		ITsconfigBuildPath[] tsconfigBuildPaths = getTypeScriptBuildPath().getTsconfigBuildPaths();
+		List<String> result = new ArrayList<>(tsconfigBuildPaths.length);
+		for (ITsconfigBuildPath tsconfigBuildPath : tsconfigBuildPaths) {
+			IFile tsconfigFile = tsconfigBuildPath.getTsconfigFile();
+			result.add(tsconfigFile.getLocation().makeRelativeTo(tsconfigFile.getProject().getLocation()).toString());
```
Comment (angelozerr) [inline eclipse/ts.eclipse.ide.core/src/ts/eclipse/ide/internal/core/resources/IDETypeScriptProject.java]: Please use WorkbenchResourceUtil.getRelativePath(tsconfigFile, tsconfigFile.getProject())

**Comment by angelozerr** [comment]
Wow great PR @lorenzodallavecchia! Thanks! I have added some little comments. Please fix it.

File: core/ts.core/src/ts/client/ITypeScriptServiceClient.java (hunk 1 above)
Comment (lorenzodallavecchia) [inline core/ts.core/src/ts/client/ITypeScriptServiceClient.java]: I had the correct preferences configured, but somehow the format on save did not include some lines. I'm double-checking that now.

**Comment by angelozerr** [comment]
Thanks a lot @lorenzodallavecchia ! Very cool PR.

**Comment by angelozerr** [comment]
@lorenzodallavecchia now when I open file, the editor is freezing because of waitForFuture. I would like really avoid freezing editor. Could you fix that please?

**Comment by angelozerr** [comment]
@lorenzodallavecchia I have uncomment the use of waitForFuture to avoid freeze of editor. Don't hesitate to do PR if you find a better solution. I would like to avoid freezing the editor.

**Comment by angelozerr** [comment]
@lorenzodallavecchia I'm really sorry. I have reverted your PR because the editor froze every time. And sometimes it freezees Eclipse IDE (I had to kill it).

I would like to do a release this month, so I would like to have a stable version. I would like to integrate your PR but not for 1.2.0 and study why there are so many freeze. Hope you will understand.

**Comment by lorenzodallavecchia** [comment]
@angelozerr don't worry: I understand.
I am perfectly fine with that feature not landing in 1.2.0, but I hope it will land eventually because the current issues with language services as well as issues #143 and #144 are preventing me from using compile on save, which is far superior to _tsc_-based build on save.

The freezing problem is really strange I did test opening and closing editors many times and experienced absolutely no freezing. Also, that waiting stuff should not actually change anything compared to now because:

- if the caller was using `CompletableFuture` (the most common case), than the caller would just be waiting for a chain of two completion stages instead of one,

- if the caller was using `get` on the future to wait for the result, it was already waiting before my change.

So, overall really really strange.
Could you provide some directions to reproduce the freezing problem?

**Comment by angelozerr** [comment]
@lorenzodallavecchia I will try to investigate more but my "main" project that I test is https://github.com/Microsoft/vscode which is big.  For instance when I open https://github.com/Microsoft/vscode/blob/master/extensions/javascript/src/features/packageJSONContribution.ts there is a little freeze. After when I navigate quicly (must find what I do exactly), Eclipse IS freeze and I must kill it.

**Comment by lorenzodallavecchia** [comment]
Hi @angelozerr, I finally had time to look into this issue again.

I tried checking out https://github.com/Microsoft/vscode into a single Eclipse project. Then I opened many files (up to 60 at the same time) and navigated back and forth using Ctrl+click and outlines.
Apart from the initial wait (which I think is due to the project being large), I experienced no freezing at all.
I'm really puzzled.

Can you give me other pointers about how to reproduce the issue?

Otherwise I think I would have to remake this PR without the CompletableFuture change: it would be more fragile but still hopefully able to fix #142.

**Comment by angelozerr** [comment]
Hi @lorenzodallavecchia ! Glad to have your news:)

To be honnest with you, I had tested quicly, but I would like really avoid having freezing when user open an editor even if tsserver is not available. Today it works like this. The only freeze that it exists today is when user open completion, hyperlink but for other features (outline, mark occurences, etc), the editor doesn't freeze.

For completion, Oxygen should provide an async completion (hope it will resolve the freeze problem). For Hyperlink, hope Eclipse guys will fix that.

I will b every glad to have a PR, but please avoid having freeze even the first time when editor is opened. Hope you will understand my wish.

I'm very busy with other topics like autoimport with completion, integrates angular language service (inside HTML) and CodeLens, so it's difficult for me to help you with your issue. Any PR are welcome!

**Comment by lorenzodallavecchia** [comment]
Thanks for helping out @angelozerr.

I completely agree with the design principle of postponing waits as much as possible.
Since I was trying with smaller projects, the "first-time" wait was not present. Now that I see that, I'm am convinced too that chaining a wait before file opening is not a wise thing to do.

I think I will investigate inside tsserver to be sure that even if `openExternalProject` finishes after `openFile`, the server state remains consistent.

**Comment by angelozerr** [comment]
@lorenzodallavecchia I think you should study too the VSCode TypeScript client. For instance https://github.com/Microsoft/vscode/blob/master/extensions/typescript/src/typescriptServiceClient.ts#L697

**Comment by lorenzodallavecchia** [comment]
Ok @angelozerr, I finally have a clear understanding of this problem.

First of all, VSCode is **not using `openExternalProject`**. That means the server is automatically picking up the `tsconfig.json` files that are closest to the edited files. Since the VSCode source is made up of many small projects, the language server is fast, since it only has to manage small projects and only a few at once (no more than the number of open files).
The downside is that there is no cross-project error checking in VSCode (exactly what I would like to have in Eclipse).

The fix I proposed in this PR was causing a slowdown when editing VSCode because it caused **all** TS projects to be loaded when opening just a single file. While this enables cross-project checking, it also wastes a lot of resources.
The situation was acceptable in my tests because I mostly use Eclipse projects containing only one `tsconfig.json`, so the cost of loading a single file is the same.

I tried removing the wait for `Future` values, but that only moved the problem to the subsequent `tsserver` calls. For example, the Outline was taking 15 seconds to appear! Ditto for the content assistant. Of course, that is not acceptable.

Unfortunately, this situation is a limitation of TypeScript and I think we would have to wait for them to implement Microsoft/TypeScript#3469. That issue will introduce a way for a `tsconfig.json` file to reference another and have cross-project checking without impacting performance.

**Comment by angelozerr** [comment]
Many thank's @lorenzodallavecchia for your great feedback!

If I have understood, we must wait for Microsoft/TypeScript#3469 to fix this issue.

**Pull request merged**